### ISSUE TITLE
Brackets not loading code folding states

### ISSUE THREAD

**Issue by derrickb** [adobe/brackets#11775]
I'm not sure exactly when this started happening, but Brackets is no longer loading my code folding states after closing/reopening.

I can see in my state.json that states are getting saved in there.  I tried deleting state.json and starting with a fresh one, but the states still aren't being loaded correctly.

I have deleted several problematic extensions in the past week, one of which was the Custom Region Code Folding extension which was giving me issues, like not showing the fold arrow icon changing when folding/unfolding.  Is it possible there's something residual from that?

The other extensions I have installed:
Autoprefixer, Brackets Outline List, Brackets Sort Text, CSSLint, Extension Rating, Minimap, Paste and Indent, Pop-up Menu Brackets, Preferences Setup UI, Visual Studio Dark theme

**Comment by derrickb** [comment]
An update: I just did a complete uninstall using Revo Uninstaller and reinstalled Brackets.  Unfortunately it is still not remember my code folding states.

**Comment by derrickb** [comment]
An update: I did a complete uninstall of Brackets including the data in AppData, and reinstalled 1.4.  Unfortunately Brackets is still not restoring my code folding states.

Closed issue by accident, reopening...

**Comment by abose** [comment]
Tagging author @thehogfather .

**Comment by thehogfather** [comment]
Are there any error messages in the console?
i.e., if you go to the menu and choose Debug > Show Developer Tools what do you see?
In addition what operating system are you running?

**Comment by derrickb** [comment]
Thanks for the reply.  OS is Windows 7 x64 Professional.

I do have errors, here they are:

* (Many times) Uncaught TypeError: Cannot read property 'from' of undefined - foldgutter.js:2
* Worker: Error thrown in tern_worker:Cannot read property 'length' of undefined - ScopeManager.js:1073
* Failed to load resource: net::ERR_NETWORK_IO_SUSPENDED - http://online-dnbard.rhcloud.com/tick/53d5549ee31c428fe600006f/84650d6fabb8009f0e44b31e171da272
* Can't track online status, retry in 5 mins - online.js:72

I'm a little concerned by this "online status" error, why is Brackets trying to connect to something?  I have health report disabled.

**Comment by anilnatha** [comment]
I am experiencing this same issue. If there is any error information I can provide from my Win7 box running Brackets ver.  1.4.0-16380 please let me know.

**Comment by thehogfather** [comment]
Thanks @derrickb for the better error reports. I've been busy all week so I couldnt look into this. 
Your online status errors are probably from the extensions rating plugin and not the health report.

@anilnatha if you have any errors please post them here. Just go to the menu and choose Debug > Show Developer Tools. It would be useful to know if the stack is similar to the one posted by @derrickb

**Comment by thehogfather** [comment]
@derrickb also if you had any problems with the custom region folding extension, would you mind filing an issue on that extension's repo so that I can investigate and fix when I get some time.

**Comment by derrickb** [comment]
@thehogfather I just followed your suggested steps and it worked! The folding states were remembered for both file 1 and file 2.

I did a couple more experiments and I think I can verify that the fold states only save when you navigate away from that file.  So the repro seems to be:

Change a fold state, and close Brackets.  State will NOT save.
Change a fold state, click on a different working file, close Brackets.  State DID save.

Also, the working file you change focus to does NOT need to be in the same pane/view. In other words if I change a fold state in the left pane, and then change focus to a working file in the right pane, and close Brackets, the left pane file's states do load correctly when I reopen Brackets.

Oh, and all of this is on another fresh install of Brackets 1.5 plus a few extensions (Minimap, outline, pop-up menu).

**Comment by thehogfather** [comment]
Thanks for the update. It confirms my suspicion that the `beforeAppClose` event  on the extension is not being fired on Windows. I'll have to check if it a brackets bug or a bug on the extension. Let's keep this open.

**Comment by xtoq** [comment]
@thehogfather should the steps you outlined on [Oct 10](https://github.com/adobe/brackets/issues/11775#issuecomment-147071686) work for plain text files?

**Comment by thehogfather** [comment]
@xtoq Yes it should if you have the `"code-folding.alwaysUseIndentFold:"` property set to `true` in your preferences.

**Comment by derrickb** [comment]
I'm still seeing this issue in 1.8 build 1.8.0-17080

**Comment by paulmz** [comment]
I've been experiencing a similar issue for a long time in current and previous builds. Using build 1.9 (1.9.0-17312 on macOS 10.12.4) I have a persistent code-fold-saving failure specifically with <style> and <script> blocks in an HTML file. All other basic HTML tag blocks save their folding states but every single time I open a file all <style> and <script> blocks are unfolded.

Is this part of the issue others are seeing or something different?

**Comment by xtoq** [comment]
@paulmz That's not what I was experiencing before. For me, I was experiencing code folding not maintaining the folds in plain text files after they had been closed. I haven't experienced this specific problem since at least v1.8 in text files, and I don't usually code fold in HTML files as I use Jekyll and WordPress so things are pretty modular. 

I did just test a few different HTML projects though, and when I folded <script> and <style> blocks, closed the files, closed Brackets, and then reopened both I still had folded code.

Have you tried reloading without extensions (Debug > Reload Without Exentions or shift+F5) and see if the problem still occurs?
